vars/npmExecuteEndToEndTests.groovy

```diff
@@ -0,0 +1,137 @@
+        }
+        if (!(config.appUrls instanceof List)) {
+            error "[${STEP_NAME}] The execution failed, since appUrls is not a list. Please provide appUrls as a list of maps."
+        }
+        if (!config.runScript) {
+            error "[${STEP_NAME}] No runScript was defined."
+        }
+
+        for (int i = 0; i < config.appUrls.size(); i++) {
+            List credentials = []
+            def appUrl = config.appUrls[i]
+
+            if (!(appUrl instanceof Map)) {
+                error "[${STEP_NAME}] The element ${appUrl} is not of type map. Please provide appUrls as a list of maps."
+            }
+            if (!appUrl.url) {
+                error "[${STEP_NAME}] No url property was defined for the following element in appUrls: ${appUrl}"
+            }
+            if (appUrl.credentialId) {
+                credentials.add([$class: 'UsernamePasswordMultiBinding', credentialsId: appUrl.credentialId, passwordVariable: 'e2e_password', usernameVariable: 'e2e_username'])
+            }
+
+            Closure e2eTest = {
+                Utils utils = new Utils()
+                utils.unstashStageFiles(script, stageName)
+                try {
+                    withCredentials(credentials) {
+                        if (appUrl.parameters) {
+                            if (appUrl.parameters instanceof List) {
+                                npmExecuteScripts(script: script, parameters: npmParameters, install: false, virtualFrameBuffer: true, runScripts: [config.runScript], scriptOptions: ["--launchUrl=${appUrl.url}", appUrl.parameters])
+                            } else {
+                                error "[${STEP_NAME}] The parameters property is not of type list. Please provide parameters as a list of strings."
+                            }
+                        }
+                        npmExecuteScripts(script: script, parameters: npmParameters, install: false, virtualFrameBuffer: true, runScripts: [config.runScript], scriptOptions: ["--launchUrl=${appUrl.url}"])
+                    }
+
+                } catch (Exception e) {
+                    error "[${STEP_NAME}] The execution failed with error: ${e.getMessage()}"
+                } finally {
+                    //TODO: Implement Report handling
+                    utils.stashStageFiles(script, parameters.stage)
+                }
+            }
+            e2ETests["E2E Tests ${index > 1 ? index : ''}"] = {
+                if (env.POD_NAME) {
+                    dockerExecuteOnKubernetes(script: script, containerMap: ContainerMap.instance.getMap().get(parameters.stage) ?: [:]) {
+                        e2eTest.call()
+                    }
+                } else {
+                    node(env.NODE_NAME) {
+                        e2eTest.call()
+                    }
+                }
+            }
+            index++
+        }
+        runClosures(e2ETests, config)
+    }
+}
+
+def runClosures(Map toRun, Map config) {
```

Comment:
maybe this should be a step now? I've originaly did it as a function in multicloud deploy, but if we need it in multiple places it is probably better to make a step

Reply:
Yes that's right, I'll move it to a separate step.